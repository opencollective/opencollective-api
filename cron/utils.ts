import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
