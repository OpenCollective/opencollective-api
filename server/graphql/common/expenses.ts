import { ExpenseAttachment } from '../../models/ExpenseAttachment';

type ViewExpenseDataPermission = {
  attachments: boolean;
  payoutMethod: boolean;
  userLocation: boolean;
  invoiceInfo: boolean;
};

/**
 * Returns true if user is allowed to see the private infos for an expense.
 * Based on loaders, so it's safe to use this function in lists.
 */
export const canViewExpensePrivateInfo = async (expense, req): Promise<ViewExpenseDataPermission> => {
  if (!req.remoteUser) {
    return {
      attachments: false,
      payoutMethod: false,
      userLocation: false,
      invoiceInfo: false,
    };
  } else if (req.remoteUser.isAdmin(expense.FromCollectiveId) || req.remoteUser.id === expense.UserId) {
    // Users can see the all the private info for the collective they're admin of and the expenses they submit
    return {
      attachments: true,
      payoutMethod: true,
      userLocation: true,
      invoiceInfo: true,
    };
  } else if (req.remoteUser.isAdmin(expense.CollectiveId)) {
    // Collective admins can see the attachments only
    return {
      attachments: true,
      payoutMethod: false,
      userLocation: false,
      invoiceInfo: false,
    };
  } else {
    // Users can see the private info for the expenses submitted to collectives they're hosting
    const collective = await req.loaders.Collective.byId.load(expense.CollectiveId);
    const isHostAdmin = req.remoteUser.isAdmin(collective.HostCollectiveId);
    const isParentCollectiveAdmin = req.remoteUser.isAdmin(collective.ParentCollectiveId);
    return {
      attachments: isHostAdmin || isParentCollectiveAdmin,
      payoutMethod: isHostAdmin,
      userLocation: isHostAdmin,
      invoiceInfo: isHostAdmin,
    };
  }
};

/**
 * Returns the list of attachments for this expense.
 */
export const getExpenseAttachments = async (expenseId, req): Promise<ExpenseAttachment[]> => {
  return req.loaders.ExpenseAttachment.byExpenseId.load(expenseId);
};
