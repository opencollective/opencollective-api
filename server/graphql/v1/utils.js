import debugLib from 'debug';
import { graphql } from 'graphql';

import { loaders } from '../loaders';

import schema from './schema';

const debug = debugLib('graphql');

export const makeRequest = (remoteUser, query) => {
  return {
    remoteUser,
    body: { query },
    loaders: loaders({ remoteUser }),
  };
};

export const graphqlQuery = async (query, variables, remoteUser) => {
  const prepare = () => {
    if (remoteUser) {
      remoteUser.rolesByCollectiveId = null; // force refetching the roles
      return remoteUser.populateRoles();
    } else {
      return Promise.resolve();
    }
  };

  if (process.env.DEBUG && process.env.DEBUG.match(/graphql/)) {
    debug('query', query);
    debug('variables', variables);
    debug('context', remoteUser);
  }

  return prepare().then(() =>
    graphql({
      schema,
      source: query,
      rootValue: null,
      contextValue: makeRequest(remoteUser, query),
      variableValues: variables,
    }),
  );
};
