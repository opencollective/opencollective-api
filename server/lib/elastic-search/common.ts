import config from 'config';

import { ElasticSearchIndexName } from './constants';

const getIndexesPrefix = () => config.elasticSearch?.indexesPrefix;

/**
 * Formats the index name before querying ElasticSearch. Allows to share a single ElasticSearch
 * instance between multiple environments (e.g. staging and production, dev and test).
 */
export const formatIndexNameForElasticSearch = (indexName: ElasticSearchIndexName): string => {
  const prefix = getIndexesPrefix();
  if (prefix) {
    return `${prefix}_${indexName}`;
  } else {
    return indexName;
  }
};
