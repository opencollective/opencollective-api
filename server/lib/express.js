import { Strategy as TwitterStrategy } from '@superfaceai/passport-twitter-oauth2';
import cloudflareIps from 'cloudflare-ip/ips.json';
import config from 'config';
import RedisStore from 'connect-redis';
import cookieParser from 'cookie-parser';
import cors from 'cors';
import errorHandler from 'errorhandler';
import express from 'express';
import session from 'express-session';
import helmet from 'helmet';
import { get, has } from 'lodash';
import passport from 'passport';
import { Strategy as GitHubStrategy } from 'passport-github';

import { loadersMiddleware } from '../graphql/loaders';

import hyperwatch from './hyperwatch';
import logger from './logger';
import { createRedisClient, RedisInstanceType } from './redis';

export default async function (app) {
  app.set('trust proxy', ['loopback', 'linklocal', 'uniquelocal'].concat(cloudflareIps));

  app.use(
    helmet({
      // It's currently breaking GraphQL playgrounds, to consider when activating this
      contentSecurityPolicy: false,
    }),
  );

  // Loaders are attached to the request to batch DB queries per request
  // It also creates in-memory caching (based on request auth);
  app.use(loadersMiddleware);

  // Body parser.
  app.use(
    express.json({
      limit: '50mb',
      // If the request is routed to our /webhooks/transferwise endpoint, we add
      // the request body buffer to a new property called `rawBody` so we can
      // calculate the checksum to verify if the request is authentic.
      verify(req, res, buf) {
        if (req.originalUrl.startsWith('/webhooks')) {
          req.rawBody = buf.toString();
        }
      },
    }),
  );
  app.use(express.urlencoded({ limit: '50mb', extended: true }));

  // Hyperwatch
  await hyperwatch(app);

  // Error handling.
  if (config.env !== 'production' && config.env !== 'staging') {
    app.use(errorHandler());
  }

  // Cors.
  app.use(cors());

  const verify = (accessToken, tokenSecret, profile, done) => done(null, accessToken, { tokenSecret, profile });

  // Github
  if (has(config, 'github.clientID') && has(config, 'github.clientSecret')) {
    passport.use(new GitHubStrategy(get(config, 'github'), verify));
  } else {
    logger.info('Configuration missing for passport GitHubStrategy, skipping.');
  }

  // Twitter
  const twitterConfig = get(config, 'twitter');
  if (has(twitterConfig, 'consumerKey') && has(twitterConfig, 'consumerSecret')) {
    passport.use(
      new TwitterStrategy(
        {
          clientType: 'confidential',
          clientID: twitterConfig.consumerKey,
          clientSecret: twitterConfig.consumerSecret,
        },
        verify,
      ),
    );
  } else {
    logger.info('Configuration missing for passport TwitterStrategy, skipping.');
  }

  app.use(cookieParser());

  // Setup session (required by passport)

  const redisClient = await createRedisClient(RedisInstanceType.SESSION);
  if (redisClient || process.env.OC_ENV === 'development') {
    const store = !redisClient ? undefined : new RedisStore({ client: redisClient });
    app.use(
      session({
        store,
        secret: config.keys.opencollective.sessionSecret,
        resave: false,
        saveUninitialized: false,
        cookie: {
          maxAge: 24 * 60 * 60 * 1000, // 1 day
          httpOnly: true,
          secure: config.env === 'production' || config.env === 'staging',
        },
      }),
    );

    app.use(passport.initialize());
  }
}
