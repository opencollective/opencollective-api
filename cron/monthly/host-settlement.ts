#!/usr/bin/env node
import '../../server/env';

import config from 'config';
import { parse as json2csv } from 'json2csv';
import { groupBy, pick, sumBy } from 'lodash';
import moment from 'moment';
import { v4 as uuid } from 'uuid';

import activityType from '../../server/constants/activities';
import expenseStatus from '../../server/constants/expense_status';
import expenseTypes from '../../server/constants/expense_type';
import { SETTLEMENT_EXPENSE_PROPERTIES } from '../../server/constants/transactions';
import { uploadToS3 } from '../../server/lib/awsS3';
import { getPendingHostFeeShare, getPendingPlatformTips } from '../../server/lib/host-metrics';
import { parseToBoolean } from '../../server/lib/utils';
import models, { sequelize } from '../../server/models';
import { PayoutMethodTypes } from '../../server/models/PayoutMethod';

const today = moment.utc();

const defaultDate = process.env.START_DATE ? moment.utc(process.env.START_DATE) : moment.utc();

const DRY = process.env.DRY;
const HOST_ID = process.env.HOST_ID;
const isProduction = config.env === 'production';

// Only run on the 1th of the month
if (isProduction && new Date().getDate() !== 1 && !process.env.OFFCYCLE) {
  console.log('OC_ENV is production and today is not the 1st of month, script aborted!');
  process.exit();
} else if (parseToBoolean(process.env.SKIP_HOST_SETTLEMENT)) {
  console.log('Skipping because SKIP_HOST_SETTLEMENT is set.');
  process.exit();
}

if (DRY) {
  console.info('Running dry, changes are not going to be persisted to the DB.');
}

const ATTACHED_CSV_COLUMNS = ['createdAt', 'description', 'amount', 'currency', 'OrderId', 'TransactionGroup'];

export async function run(baseDate: Date | moment.Moment = defaultDate): Promise<void> {
  const momentDate = moment(baseDate).subtract(1, 'month');
  const year = momentDate.year();
  const month = momentDate.month();
  const startDate = new Date(year, month, 1);
  const endDate = new Date(year, month + 1, 1);

  console.info(`Invoicing hosts pending fees and tips for ${momentDate.format('MMMM')}.`);

  const payoutMethods = groupBy(
    await models.PayoutMethod.findAll({
      where: { CollectiveId: SETTLEMENT_EXPENSE_PROPERTIES.FromCollectiveId, isSaved: true },
    }),
    'type',
  );

  const hosts = await sequelize.query(
    `
      SELECT c.*
      FROM "Collectives" c
      INNER JOIN "Transactions" t ON t."HostCollectiveId" = c.id AND t."deletedAt" IS NULL
      WHERE c."isHostAccount" IS TRUE
      AND t."createdAt" >= :startDate AND t."createdAt" < :endDate
      AND c.id != 8686 -- Make sure we don't invoice OC Inc as reverse settlements are not supported yet
      GROUP BY c.id
    `,
    {
      mapToModel: true,
      type: sequelize.QueryTypes.SELECT,
      model: models.Collective,
      replacements: { startDate: startDate, endDate: endDate },
    },
  );

  for (const host of hosts) {
    const pendingPlatformTips = await getPendingPlatformTips(host, { startDate, endDate });
    const pendingHostFeeShare = await getPendingHostFeeShare(host, { startDate, endDate });

    if (HOST_ID && host.id !== parseInt(HOST_ID)) {
      continue;
    }

    const plan = await host.getPlan();

    let items = [];

    const transactions = await sequelize.query(
      `SELECT t.*
FROM "Transactions" as t
INNER JOIN "TransactionSettlements" ts ON ts."TransactionGroup" = t."TransactionGroup" AND t.kind = ts.kind
WHERE t."CollectiveId" = :CollectiveId
AND t."createdAt" >= :startDate AND t."createdAt" < :endDate
AND t."kind" IN ('PLATFORM_TIP_DEBT', 'HOST_FEE_SHARE_DEBT')
AND t."isDebt" IS TRUE
AND t."deletedAt" IS NULL
AND ts."status" != 'SETTLED'`,
      {
        replacements: { CollectiveId: host.id, startDate: startDate, endDate: endDate },
        model: models.Transaction,
        mapToModel: true, // pass true here if you have any mapped fields
      },
    );

    items.push({
      incurredAt: new Date(),
      amount: pendingPlatformTips,
      description: 'Platform Tips',
    });

    items.push({
      incurredAt: new Date(),
      amount: pendingHostFeeShare,
      description: 'Shared Revenue',
    });

    if (plan.pricePerCollective) {
      const activeHostedCollectives = await host.getHostedCollectivesCount();
      const amount = (activeHostedCollectives || 0) * plan.pricePerCollective;
      if (amount) {
        items.push({
          incurredAt: new Date(),
          amount,
          description: 'Fixed Fee per Hosted Collective',
        });
      }
    }

    const totalAmountCharged = sumBy(items, 'amount');

    if (totalAmountCharged < 1000) {
      console.warn(
        `${host.name} (#${host.id}) skipped, total amount pending ${totalAmountCharged / 100} < 10.00 ${
          host.currency
        }.\n`,
      );
      continue;
    }
    console.info(
      `${host.name} (#${host.id}) has ${transactions.length} pending transactions and owes ${
        totalAmountCharged / 100
      } (${host.currency})`,
    );

    let csv;
    if (transactions.length) {
      csv = json2csv(transactions.map(t => pick(t, ATTACHED_CSV_COLUMNS)));
    }

    if (DRY) {
      console.debug(`Items:\n${json2csv(items)}\n`);
    } else {
      const connectedAccounts = await host.getConnectedAccounts({
        where: { deletedAt: null },
      });

      let payoutMethod =
        payoutMethods[PayoutMethodTypes.OTHER]?.[0] || payoutMethods[PayoutMethodTypes.BANK_ACCOUNT]?.[0];
      if (
        connectedAccounts?.find(c => c.service === 'transferwise') &&
        payoutMethods[PayoutMethodTypes.BANK_ACCOUNT]?.[0]
      ) {
        const currencyCompatibleAccount = payoutMethods[PayoutMethodTypes.BANK_ACCOUNT].find(
          pm => pm.data?.['currency'] === host.currency,
        );
        payoutMethod = currencyCompatibleAccount || payoutMethods[PayoutMethodTypes.BANK_ACCOUNT]?.[0];
      } else if (
        connectedAccounts?.find(c => c.service === 'paypal') &&
        !host.settings?.disablePaypalPayouts &&
        payoutMethods[PayoutMethodTypes.PAYPAL]?.[0]
      ) {
        payoutMethod = payoutMethods[PayoutMethodTypes.PAYPAL]?.[0];
      }

      if (!payoutMethod) {
        console.error('No Payout Method found, Open Collective Inc. needs to have at least one payout method.');
        process.exit();
      }

      // Create the Expense
      const transactionIds = transactions.map(t => t.TransactionId);
      const expense = await models.Expense.create({
        ...SETTLEMENT_EXPENSE_PROPERTIES,
        PayoutMethodId: payoutMethod.id,
        amount: totalAmountCharged,
        CollectiveId: host.id,
        currency: host.currency,
        description: `Platform settlement for ${momentDate.utc().format('MMMM')}`,
        incurredAt: today,
        data: { isPlatformTipSettlement: true, transactionIds },
        type: expenseTypes.INVOICE,
        status: expenseStatus.PENDING,
      });

      // Create Expense Items
      items = items.map(i => ({
        ...i,
        ExpenseId: expense.id,
        CreatedByUserId: SETTLEMENT_EXPENSE_PROPERTIES.UserId,
      }));

      await models.ExpenseItem.bulkCreate(items);

      // Attach CSV
      if (csv) {
        const Body = csv;
        const filenameBase = `${host.name}-${momentDate.format('MMMM-YYYY')}`;
        const Key = `${filenameBase}.${uuid().split('-')[0]}.csv`;
        const { Location: url } = await uploadToS3({
          Bucket: config.aws.s3.bucket,
          Key,
          Body,
          ACL: 'public-read',
          ContentType: 'text/csv',
        });
        await models.ExpenseAttachedFile.create({
          url,
          ExpenseId: expense.id,
          CreatedByUserId: SETTLEMENT_EXPENSE_PROPERTIES.UserId,
        });
      }

      // Mark transactions as invoiced
      await models.TransactionSettlement.markTransactionsAsInvoiced(transactions, expense.id);
      await expense.createActivity(activityType.COLLECTIVE_EXPENSE_CREATED);
    }
  }
}

if (require.main === module) {
  run(defaultDate)
    .catch(e => {
      console.error(e);
      process.exit(1);
    })
    .then(() => {
      process.exit();
    });
}
