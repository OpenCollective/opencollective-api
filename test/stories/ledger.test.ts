/**
 * This test is meant to test the common workflows for the ledger: record a contribution,
 * refund it, add platform tips, etc.
 */

import { expect } from 'chai';
import express from 'express';
import moment from 'moment';

import { run as runSettlementScript } from '../../cron/monthly/host-settlement';
import {
  PLATFORM_TIP_TRANSACTION_PROPERTIES,
  SETTLEMENT_EXPENSE_PROPERTIES,
} from '../../server/constants/transactions';
import { payExpense } from '../../server/graphql/common/expenses';
import { createRefundTransaction, executeOrder } from '../../server/lib/payments';
import models from '../../server/models';
import { fakeCollective, fakeHost, fakeOrder, fakePayoutMethod, fakeUser } from '../test-helpers/fake-data';
import { resetTestDB, snapshotLedger } from '../utils';

const SNAPSHOT_COLUMNS = [
  'kind',
  'type',
  'amount',
  'paymentProcessorFeeInHostCurrency',
  'CollectiveId',
  'FromCollectiveId',
  'HostCollectiveId',
  'settlementStatus',
  'isRefund',
];

describe('test/stories/ledger', () => {
  let collective, host, hostAdmin, ocInc, contributorUser, baseOrderData;

  beforeEach(async () => {
    // TODO: The setup should ideally insert other hosts and transactions to make sure the balance queries are filtering correctly

    await resetTestDB();
    hostAdmin = await fakeUser();
    host = await fakeHost({ name: 'OSC', admin: hostAdmin.collective });
    await hostAdmin.populateRoles();
    await host.update({ HostCollectiveId: host.id, isActive: true });
    collective = await fakeCollective({ HostCollectiveId: host.id, name: 'ESLint', hostFeePercent: 5 });
    contributorUser = await fakeUser(undefined, { name: 'Ben' });
    ocInc = await fakeHost({ name: 'OC Inc', id: PLATFORM_TIP_TRANSACTION_PROPERTIES.CollectiveId });
    await fakePayoutMethod({ type: 'OTHER', CollectiveId: ocInc.id }); // For the settlement expense
    await fakeUser({ id: SETTLEMENT_EXPENSE_PROPERTIES.UserId, name: 'Pia' });
    baseOrderData = {
      description: `Financial contribution to ${collective.name}`,
      totalAmount: 10000,
      FromCollectiveId: contributorUser.CollectiveId,
      CollectiveId: collective.id,
      PaymentMethodId: null,
    };
  });

  afterEach(async () => {
    const transactions = await models.Transaction.findAll({ order: [['id', 'DESC']] });
    await Promise.all(
      transactions.map(transaction => models.Transaction.validate(transaction, { validateOppositeTransaction: true })),
    );
  });

  describe('Level 1: Same currency', () => {
    it('1. Simple contribution without host fees', async () => {
      await collective.update({ hostFeePercent: 0 });
      const order = await fakeOrder(baseOrderData);
      order.paymentMethod = { service: 'opencollective', type: 'manual', paid: true };
      await executeOrder(contributorUser, order);

      await snapshotLedger(SNAPSHOT_COLUMNS);
      expect(await collective.getBalance()).to.eq(10000);
      expect(await host.getTotalMoneyManaged()).to.eq(10000);
      expect(await host.getBalance()).to.eq(0);
      expect(await ocInc.getBalance()).to.eq(0);
    });

    it('2. Simple contribution with 5% host fees', async () => {
      const order = await fakeOrder(baseOrderData);
      order.paymentMethod = { service: 'opencollective', type: 'manual', paid: true };
      await executeOrder(contributorUser, order);

      await snapshotLedger(SNAPSHOT_COLUMNS);
      expect(await collective.getBalance()).to.eq(9500); // 1000 - 5% host fee
      expect(await host.getTotalMoneyManaged()).to.eq(10000);
      expect(await host.getBalance()).to.eq(500); // 5% host fee
      expect(await ocInc.getBalance()).to.eq(0);
    });

    it('3. Simple contribution with 5% host fees and indirect platform tip (unsettled)', async () => {
      const order = await fakeOrder({ ...baseOrderData, data: { isFeesOnTop: true, platformFee: 1000 } });
      order.paymentMethod = { service: 'opencollective', type: 'manual', paid: true };
      await executeOrder(contributorUser, order);

      await snapshotLedger(SNAPSHOT_COLUMNS);
      expect(await collective.getBalance()).to.eq(8550); // (10000 Total - 1000 platform tip) - 5% host fee (450)
      expect(await host.getTotalMoneyManaged()).to.eq(10000); // Tip is still on host's account
      expect(await host.getBalance()).to.eq(1450);
      expect(await host.getBalanceWithBlockedFunds()).to.eq(1450);
      // TODO We should have a "Projected balance" that removes everything owed
      expect(await ocInc.getBalance()).to.eq(0);
    });

    it('4. Simple contribution with 5% host fees and indirect platform tip (settled)', async () => {
      // Create initial order
      const order = await fakeOrder({ ...baseOrderData, data: { isFeesOnTop: true, platformFee: 1000 } });
      order.paymentMethod = { service: 'opencollective', type: 'manual', paid: true };
      await executeOrder(contributorUser, order);

      // Run host settlement
      await runSettlementScript(moment().add(1, 'month').toDate());
      const settlementExpense = await models.Expense.findOne();
      expect(settlementExpense).to.exist;
      await settlementExpense.update({ status: 'APPROVED' });
      await payExpense(<express.Request>{ remoteUser: hostAdmin }, { id: settlementExpense.id, forceManual: true });

      // Check data
      await snapshotLedger(SNAPSHOT_COLUMNS);
      expect(await collective.getBalance()).to.eq(8550); // (10000 Total - 1000 platform tip) - 5% host fee (450)
      expect(await host.getTotalMoneyManaged()).to.eq(9000); // 10000 - 1000, platform tip is not there anymore
      expect(await host.getBalance()).to.eq(450);
      expect(await host.getBalanceWithBlockedFunds()).to.eq(450);
      expect(await ocInc.getBalance()).to.eq(1000);
      expect(await ocInc.getBalanceWithBlockedFunds()).to.eq(1000);
    });

    it('5. Refunded contribution with host fees, payment processor fees and indirect platform tip', async () => {
      // Create initial order
      const order = await fakeOrder({
        ...baseOrderData,
        data: { isFeesOnTop: true, platformFee: 1000, paymentProcessorFeeInHostCurrency: 200 },
      });
      order.paymentMethod = { service: 'opencollective', type: 'manual', paid: true };
      await executeOrder(contributorUser, order);

      // Run host settlement
      await runSettlementScript(moment().add(1, 'month').toDate());
      const settlementExpense = await models.Expense.findOne();
      expect(settlementExpense).to.exist;
      await settlementExpense.update({ status: 'APPROVED' });
      await payExpense(<express.Request>{ remoteUser: hostAdmin }, { id: settlementExpense.id, forceManual: true });

      // New checks for payment processor fees
      expect(await collective.getBalance()).to.eq(8350); // (10000 Total - 1000 platform tip) - 5% host fee (450) - 200 processor fees
      expect(await host.getTotalMoneyManaged()).to.eq(8800); // 10000 - 1000 - 200

      // Check host metrics pre-refund
      let hostMetrics = await host.getHostMetrics();
      expect(hostMetrics).to.deep.equal({
        hostFees: 450,
        platformFees: 0,
        pendingPlatformFees: 0,
        platformTips: 1000,
        pendingPlatformTips: 0, // Already settled
        hostFeeShare: 0,
        pendingHostFeeShare: 0,
        hostFeeSharePercent: 0,
        settledHostFeeShare: 0,
        totalMoneyManaged: 8800,
      });

      // ---- Refund transaction -----
      const contributionTransaction = await models.Transaction.findOne({
        where: { OrderId: order.id, kind: 'CONTRIBUTION', type: 'CREDIT' },
      });

      await createRefundTransaction(contributionTransaction, 0, null, null);

      // Check data
      await snapshotLedger(SNAPSHOT_COLUMNS);
      expect(await collective.getBalance()).to.eq(0);
      expect(await host.getTotalMoneyManaged()).to.eq(-1200);
      expect(await host.getBalance()).to.eq(-1200); // Will be -200 after settlement (platform tip)
      expect(await host.getBalanceWithBlockedFunds()).to.eq(-1200);
      expect(await ocInc.getBalance()).to.eq(1000);
      expect(await ocInc.getBalanceWithBlockedFunds()).to.eq(1000);

      // Check host metrics
      hostMetrics = await host.getHostMetrics();
      expect(hostMetrics).to.deep.equal({
        hostFees: 0,
        platformFees: 0,
        pendingPlatformFees: 0,
        platformTips: 0, // There was a 1000 tip, but it was refunded
        pendingPlatformTips: -1000,
        hostFeeShare: 0,
        pendingHostFeeShare: 0,
        hostFeeSharePercent: 0,
        settledHostFeeShare: 0,
        totalMoneyManaged: -1200,
      });

      // Run OC settlement
      // TODO: We should run the opposite settlement and check amount
    });
  });

  describe('Level 2: Host with different currency', () => {
    // TODO: Add multi-currencies
  });

  describe('Level 3: Host and collective with different currency', () => {
    // TODO: Add multi-currencies
  });
});
