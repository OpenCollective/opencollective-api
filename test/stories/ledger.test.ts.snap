// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/stories/ledger Level 1: Same currency 1. Simple contribution without host fees 1`] = `
"
| kind         | type   | amount | paymentFee | To     | From   | Host | Settlement | isRefund |
| ------------ | ------ | ------ | ---------- | ------ | ------ | ---- | ---------- | -------- |
| CONTRIBUTION | CREDIT | 10000  | 0          | ESLint | Ben    | OSC  |            | false    |
| CONTRIBUTION | DEBIT  | -10000 | 0          | Ben    | ESLint | NULL |            | false    |"
`;

exports[`test/stories/ledger Level 1: Same currency 2. Simple contribution with 5% host fees 1`] = `
"
| kind         | type   | amount | paymentFee | To     | From   | Host | Settlement | isRefund |
| ------------ | ------ | ------ | ---------- | ------ | ------ | ---- | ---------- | -------- |
| CONTRIBUTION | CREDIT | 10000  | 0          | ESLint | Ben    | OSC  |            | false    |
| CONTRIBUTION | DEBIT  | -10000 | 0          | Ben    | ESLint | NULL |            | false    |
| HOST_FEE     | CREDIT | 500    | 0          | OSC    | ESLint | OSC  |            | false    |
| HOST_FEE     | DEBIT  | -500   | 0          | ESLint | OSC    | OSC  |            | false    |"
`;

exports[`test/stories/ledger Level 1: Same currency 3. Simple contribution with 5% host fees and indirect platform tip (unsettled) 1`] = `
"
| kind              | type   | amount | paymentFee | To     | From   | Host   | Settlement | isRefund |
| ----------------- | ------ | ------ | ---------- | ------ | ------ | ------ | ---------- | -------- |
| CONTRIBUTION      | CREDIT | 9000   | 0          | ESLint | Ben    | OSC    |            | false    |
| CONTRIBUTION      | DEBIT  | -9000  | 0          | Ben    | ESLint | NULL   |            | false    |
| HOST_FEE          | CREDIT | 450    | 0          | OSC    | ESLint | OSC    |            | false    |
| HOST_FEE          | DEBIT  | -450   | 0          | ESLint | OSC    | OSC    |            | false    |
| PLATFORM_TIP_DEBT | CREDIT | 1000   | 0          | OSC    | OC Inc | OSC    | OWED       | false    |
| PLATFORM_TIP_DEBT | DEBIT  | -1000  | 0          | OC Inc | OSC    | OC Inc | OWED       | false    |
| PLATFORM_TIP      | CREDIT | 1000   | 0          | OC Inc | Ben    | OC Inc |            | false    |
| PLATFORM_TIP      | DEBIT  | -1000  | 0          | Ben    | OC Inc | NULL   |            | false    |"
`;

exports[`test/stories/ledger Level 1: Same currency 4. Simple contribution with 5% host fees and indirect platform tip (settled) 1`] = `
"
| kind              | type   | amount | paymentFee | To     | From   | Host   | Settlement | isRefund |
| ----------------- | ------ | ------ | ---------- | ------ | ------ | ------ | ---------- | -------- |
| EXPENSE           | CREDIT | 1000   | 0          | OC Inc | OSC    | OC Inc |            | false    |
| EXPENSE           | DEBIT  | -1000  | 0          | OSC    | OC Inc | OSC    |            | false    |
| CONTRIBUTION      | CREDIT | 9000   | 0          | ESLint | Ben    | OSC    |            | false    |
| CONTRIBUTION      | DEBIT  | -9000  | 0          | Ben    | ESLint | NULL   |            | false    |
| HOST_FEE          | CREDIT | 450    | 0          | OSC    | ESLint | OSC    |            | false    |
| HOST_FEE          | DEBIT  | -450   | 0          | ESLint | OSC    | OSC    |            | false    |
| PLATFORM_TIP_DEBT | CREDIT | 1000   | 0          | OSC    | OC Inc | OSC    | SETTLED    | false    |
| PLATFORM_TIP_DEBT | DEBIT  | -1000  | 0          | OC Inc | OSC    | OC Inc | SETTLED    | false    |
| PLATFORM_TIP      | CREDIT | 1000   | 0          | OC Inc | Ben    | OC Inc |            | false    |
| PLATFORM_TIP      | DEBIT  | -1000  | 0          | Ben    | OC Inc | NULL   |            | false    |"
`;

exports[`test/stories/ledger Level 1: Same currency 5. Refunded contribution with host fees, payment processor fees and indirect platform tip 1`] = `
"
| kind                  | type   | amount | paymentFee | To     | From   | Host   | Settlement | isRefund |
| --------------------- | ------ | ------ | ---------- | ------ | ------ | ------ | ---------- | -------- |
| CONTRIBUTION          | CREDIT | 9000   | 0          | Ben    | ESLint | NULL   |            | true     |
| CONTRIBUTION          | DEBIT  | -9000  | 0          | ESLint | Ben    | OSC    |            | true     |
| HOST_FEE              | CREDIT | 450    | 0          | ESLint | OSC    | OSC    |            | true     |
| HOST_FEE              | DEBIT  | -450   | 0          | OSC    | ESLint | OSC    |            | true     |
| PAYMENT_PROCESSOR_FEE | CREDIT | 200    | 0          | ESLint | OSC    | OSC    |            | true     |
| PAYMENT_PROCESSOR_FEE | DEBIT  | -200   | 0          | OSC    | ESLint | OSC    |            | true     |
| PLATFORM_TIP_DEBT     | CREDIT | 1000   | 0          | OC Inc | OSC    | OC Inc | OWED       | true     |
| PLATFORM_TIP_DEBT     | DEBIT  | -1000  | 0          | OSC    | OC Inc | OSC    | OWED       | true     |
| PLATFORM_TIP          | CREDIT | 1000   | 0          | Ben    | OC Inc | NULL   |            | true     |
| PLATFORM_TIP          | DEBIT  | -1000  | 0          | OC Inc | Ben    | OC Inc |            | true     |
| EXPENSE               | CREDIT | 1000   | 0          | OC Inc | OSC    | OC Inc |            | false    |
| EXPENSE               | DEBIT  | -1000  | 0          | OSC    | OC Inc | OSC    |            | false    |
| CONTRIBUTION          | CREDIT | 9000   | -200       | ESLint | Ben    | OSC    |            | false    |
| CONTRIBUTION          | DEBIT  | -8800  | -200       | Ben    | ESLint | NULL   |            | false    |
| HOST_FEE              | CREDIT | 450    | 0          | OSC    | ESLint | OSC    |            | false    |
| HOST_FEE              | DEBIT  | -450   | 0          | ESLint | OSC    | OSC    |            | false    |
| PLATFORM_TIP_DEBT     | CREDIT | 1000   | 0          | OSC    | OC Inc | OSC    | SETTLED    | false    |
| PLATFORM_TIP_DEBT     | DEBIT  | -1000  | 0          | OC Inc | OSC    | OC Inc | SETTLED    | false    |
| PLATFORM_TIP          | CREDIT | 1000   | 0          | OC Inc | Ben    | OC Inc |            | false    |
| PLATFORM_TIP          | DEBIT  | -1000  | 0          | Ben    | OC Inc | NULL   |            | false    |"
`;

exports[`test/stories/ledger Same currency 1. Simple contribution without host fees 1`] = `
"
| Group     | kind         | type   | amount | paymentFee | To     | From   | Host | Settlement |
| --------- | ------------ | ------ | ------ | ---------- | ------ | ------ | ---- | ---------- |
| #e356de2d | CONTRIBUTION | CREDIT | 10000  | 0          | ESLint | Ben    | OSC  |            |
| #e356de2d | CONTRIBUTION | DEBIT  | -10000 | 0          | Ben    | ESLint | NULL |            |"
`;

exports[`test/stories/ledger Same currency 2. Simple contribution with 5% host fees 1`] = `
"
| Group     | kind         | type   | amount | paymentFee | To     | From   | Host | Settlement |
| --------- | ------------ | ------ | ------ | ---------- | ------ | ------ | ---- | ---------- |
| #5ff054a8 | CONTRIBUTION | CREDIT | 10000  | 0          | ESLint | Ben    | OSC  |            |
| #5ff054a8 | CONTRIBUTION | DEBIT  | -10000 | 0          | Ben    | ESLint | NULL |            |
| #5ff054a8 | HOST_FEE     | CREDIT | 500    | 0          | OSC    | ESLint | OSC  |            |
| #5ff054a8 | HOST_FEE     | DEBIT  | -500   | 0          | ESLint | OSC    | OSC  |            |"
`;

exports[`test/stories/ledger Same currency 3. Simple contribution with 5% host fees and platform tip (unsettled) 1`] = `
"
| Group     | kind         | type   | amount | paymentFee | To     | From   | Host   | Settlement |
| --------- | ------------ | ------ | ------ | ---------- | ------ | ------ | ------ | ---------- |
| #82d413b2 | CONTRIBUTION | CREDIT | 9000   | 0          | ESLint | Ben    | OSC    |            |
| #82d413b2 | CONTRIBUTION | DEBIT  | -9000  | 0          | Ben    | ESLint | NULL   |            |
| #82d413b2 | HOST_FEE     | CREDIT | 450    | 0          | OSC    | ESLint | OSC    |            |
| #82d413b2 | HOST_FEE     | DEBIT  | -450   | 0          | ESLint | OSC    | OSC    |            |
| #82d413b2 | PLATFORM_TIP | CREDIT | 1000   | 0          | OSC    | OC Inc | OSC    | OWED       |
| #82d413b2 | PLATFORM_TIP | DEBIT  | -1000  | 0          | OC Inc | OSC    | OC Inc | OWED       |
| #82d413b2 | PLATFORM_TIP | CREDIT | 1000   | 0          | OC Inc | Ben    | OC Inc |            |
| #82d413b2 | PLATFORM_TIP | DEBIT  | -1000  | 0          | Ben    | OC Inc | NULL   |            |"
`;
